File: 2300-successful-pairs-of-spells-and-potions/2300-successful-pairs-of-spells-and-potions.ts
function successfulPairs(spells: number[], potions: number[], success: number): number[] {
  let ans = [],
      N = spells.length,
      M = potions.length;
  
  potions.sort((a, b) => a - b);
  
  for(let i = 0; i < N; i++) {
    
    let start = 0,
        end = M;
    while(start < end) {
      const mid = start + ((end - start) >>> 1);
      if(spells[i] * potions[mid] >= success)
        end = mid;
      else
        start = mid + 1;
    }
    
    ans.push(M - start);
  }
  
  return ans;
};

File: 2300-successful-pairs-of-spells-and-potions/2300-successful-pairs-of-spells-and-potions.java
class Solution {
    public int[] successfulPairs(int[] spells, int[] potions, long success) {
        int N = spells.length,
            M = potions.length;
        int[] ans = new int[N];
        
        Arrays.sort(potions);
        
        for(int i = 0; i < N; i++) {
            ans[i] = binarySearch(potions, spells[i], success);
        }
        
        return ans;
    }
    
    private int binarySearch(int[] arr, int x, long k) {
        int N = arr.length;
        int l = 0, h = N - 1;
        
        while(l <= h) {
            int m = l + ((h - l) / 2);
            
            if(((long)arr[m] * (long)x) < k) {
                l = m + 1;
            } else {
                h = m - 1;
            }
        }
        
        return N - l;
    }
}